### DISCUSSION TITLE
Cannot decode JWT token: Signature verification failed

### DISCUSSION THREAD

**Discussion by ChristianF88** [ckan/ckan#8039] [Q&A]
Hi,

I am using a dockerized version of `ckan 2.9.9` with `ckanext-envvars`.

When posting to the API, with a freshly generated token for a `sysadmin` user I get this error:
```
2024-01-25 13:08:14,776 ERROR [ckan.lib.api_token] Cannot decode JWT token: Signature verification failed
```

I have tried posting via `curl`, but also tried `ckanapi` - same issue. I also tried different ways of generating the token (via website, via cli), it did not make a difference.

The JWT encode and decode secrets are set via `string:`.

I **can** run **`package_search`**  but for example **`package_show` fails** and other things like **`package_create` fails** too.

Any suggestions?

**Comment by ChristianF88** [comment]
I have added an API key for testing, everything works with an API key.

**Comment by blagojabozinovski** [comment]
@ChristianF88  Is the value for the api_token in the production.ini file the same for both encode and decode secret when You create and test the api token? 
For example
api_token.jwt.encode.secret = string:ABC123456543
api_token.jwt.decode.secret = string:ABC123456543

**Comment by ChristianF88** [comment]
Hi @blagojabozinovski,

The strings for encoding and decoding are different:
```bash
CKAN___API_TOKEN__JWT__ENCODE__SECRET=string:712408sdghdavjcro3zg43o543
CKAN___API_TOKEN__JWT__DECODE__SECRET=string:ynccxbasdgeiuwqz538426432safh
```
They're defined via **ckanext-envs**.

**Comment by blagojabozinovski** [comment]
@ChristianF88 then I think that is the issue. The value for encode secret when the token was created should be the same used when You decode (and use) that token.

**Comment by ChristianF88** [comment]
@blagojabozinovski thanks for responding so quickly. I will test this later and share my findings here. However if both secrets should be the same, why do two strings have to be defined? I have not looked into the JWT algortihm but that seems a bit weird to me.

**Comment by ChristianF88** [comment]
I tested with encode and decode strings being the same and it does work perfectly, thanks @blagojabozinovski. I assume there are two fields available because of the other cryptographic algorithms available for token handling.  In my opinion it would be helpful to improve the error message, or even run some kind of validation, if the settings make sense for the selected cryptographic algorithm.

**Comment by david-bernabeu** [comment]
Sorry but i've read what you have done and for me it doesn't work. I get this logs in mi CKAN:

`
2025-02-13 09:24:47,249 ERROR [ckan.lib.api_token] Cannot decode JWT token: Signature verification failed
2025-02-13 09:24:47,252 INFO  [ckan.config.middleware.flask_app]  403 /api/3/action/resource_show render time 0.005 seconds
2025-02-13 09:24:47,267 ERROR [ckan.lib.api_token] Cannot decode JWT token: Signature verification failed
`

and my prodicton.ini i have this:
`
api_token.jwt.encode.secret=string:zIfw6mXXNUH4DUbKtnPPnTeaijgOMEE9BUYHZVYqxyS
api_token.jwt.decode.secret=string:zIfw6mXXNUH4DUbKtnPPnTeaijgOMEE9BUYHZVYqxyS
`

@blagojabozinovski @ChristianF88

**Reply by david-bernabeu** [reply]
It occurs when i upload a file and the datapusher and the ckan try to interact

**Comment by Validooo** [comment]
@david-bernabeu  Please first check which config file your app is using — whether /srv/app/ckan.ini, /app/production.ini, or another config file.

From CKAN 2.9 onwards, the default location is /etc/ckan/default/ckan.ini.
For previous releases, the files were generally located at /etc/ckan/default/development.ini or /etc/ckan/default/production.ini.

So it depends on which CKAN version you are using.

After that, set the values of api_token.jwt.encode.secret and api_token.jwt.decode.secret to be the same. Also set beaker.session.secret (make sure it is different from api_token.jwt.decode.secret).

You can try the following example values:
```
api_token.jwt.encode.secret = string:jQ13MakIua9SB0KHPicl46YHFhKd5RZkyXcPmLFa98c
api_token.jwt.decode.secret = string:jQ13MakIua9SB0KHPicl46YHFhKd5RZkyXcPmLFa98c
beaker.session.secret = NFs8Q5h9NnNSnMTBkTttfyHJuCSoo9KRiOl6YUqEMnQ
```


The issue occurs when you change these values in production.ini, but that file is not actually being used as the active config file — in that case, your changes will have no effect. You can see which config file is being used in the first lines of the CKAN logs when it starts.
